File: frontend/src/api/reportApi.ts
import type {
  HeatmapReport,
  Period,
  QuadrantReport,
  StatusCounts,
  SummaryReport,
  TrendReport,
} from '../types/report'
import { api } from './client'
import { currentZone } from './datetime'
import { taskApi } from './taskApi'

const BASE = '/api/reports'

export const reportApi = {
  summary(period: Period, signal?: AbortSignal): Promise<SummaryReport> {
    return api.get<SummaryReport>(`${BASE}/summary`, { period, zone: currentZone() }, signal)
  },

  trend(period: Period, signal?: AbortSignal): Promise<TrendReport> {
    return api.get<TrendReport>(`${BASE}/trend`, { period, zone: currentZone() }, signal)
  },

  quadrants(period: Period, signal?: AbortSignal): Promise<QuadrantReport> {
    return api.get<QuadrantReport>(`${BASE}/quadrants`, { period, zone: currentZone() }, signal)
  },

  /** Open tasks as of now: takes no period and no zone. */
  distribution(signal?: AbortSignal): Promise<QuadrantReport> {
    return api.get<QuadrantReport>(`${BASE}/distribution`, undefined, signal)
  },

  heatmap(year: number, signal?: AbortSignal): Promise<HeatmapReport> {
    return api.get<HeatmapReport>(`${BASE}/heatmap`, { year, zone: currentZone() }, signal)
  },

  async statusCounts(signal?: AbortSignal): Promise<StatusCounts> {
    const [todo, inProgress, done] = await Promise.all([
      taskApi.countByStatus('TODO', signal),
      taskApi.countByStatus('IN_PROGRESS', signal),
      taskApi.countByStatus('DONE', signal),
    ])
    return { TODO: todo, IN_PROGRESS: inProgress, DONE: done }
  },
}

File: mobile/lib/api/report_api.dart
import '../models/report.dart';
import '../models/task.dart';
import 'api_client.dart';
import 'zone.dart';
import 'task_api.dart';

class ReportApi {
  ReportApi(this._client, this._tasks);
  final ApiClient _client;
  final TaskApi _tasks;


  Future<SummaryReport> summary(Period period) async => SummaryReport.fromJson(
      await _client.request('GET', '/api/reports/summary',
          query: {'period': periodWire[period]!, 'zone': deviceZone}) as Map<String, dynamic>);

  Future<TrendReport> trend(Period period) async => TrendReport.fromJson(
      await _client.request('GET', '/api/reports/trend',
          query: {'period': periodWire[period]!, 'zone': deviceZone}) as Map<String, dynamic>);

  Future<QuadrantReport> quadrants(Period period) async => QuadrantReport.fromJson(
      await _client.request('GET', '/api/reports/quadrants',
          query: {'period': periodWire[period]!, 'zone': deviceZone}) as Map<String, dynamic>);

  Future<QuadrantReport> distribution() async => QuadrantReport.fromJson(
      await _client.request('GET', '/api/reports/distribution') as Map<String, dynamic>);

  Future<StatusCounts> statusCounts() async {
    // independent calls, so they go out together rather than in sequence
    final counts = await Future.wait([
      _tasks.countByStatus(TaskStatus.todo),
      _tasks.countByStatus(TaskStatus.inProgress),
      _tasks.countByStatus(TaskStatus.done),
    ]);
    return StatusCounts(counts[0], counts[1], counts[2]);
  }
}
